src/stream/xml_reader.rs

```diff
@@ -1,219 +1,205 @@
+use std::io::{self, BufReader, Read};
 
 use crate::{
     error::{Error, ErrorKind, FilePosition},
     stream::{Event, OwnedEvent},
     Date, Integer,
 };
 
+#[derive(Clone, PartialEq, Eq)]
+struct ElmName(Box<[u8]>);
+
+impl From<&[u8]> for ElmName {
+    fn from(bytes: &[u8]) -> Self {
+        ElmName(Box::from(bytes))
+    }
+}
+
+impl AsRef<[u8]> for ElmName {
+    fn as_ref(&self) -> &[u8] {
+        &*self.0
+    }
+}
+
 pub struct XmlReader<R: Read> {
-    xml_reader: EventReader<R>,
-    queued_event: Option<XmlEvent>,
-    element_stack: Vec<String>,
+    buffer: Vec<u8>,
     finished: bool,
+    state: ReaderState<R>,
 }
 
+struct ReaderState<R: Read>(EventReader<BufReader<R>>);
+
 impl<R: Read> XmlReader<R> {
     pub fn new(reader: R) -> XmlReader<R> {
-        let config = ParserConfig::new()
-            .trim_whitespace(false)
-            .whitespace_to_characters(true)
-            .cdata_to_characters(true)
-            .ignore_comments(true)
-            .coalesce_characters(true);
+        let mut xml_reader = EventReader::from_reader(BufReader::new(reader));
+        xml_reader.trim_text(true);
+        xml_reader.check_end_names(true);
+        xml_reader.expand_empty_elements(true);
 
         XmlReader {
-            xml_reader: EventReader::new_with_config(reader, config),
-            queued_event: None,
-            element_stack: Vec::new(),
+            buffer: Vec::new(),
             finished: false,
+            state: ReaderState(xml_reader),
         }
     }
+}
 
-    fn read_content(&mut self) -> Result<String, Error> {
-        loop {
-            match self.xml_reader.next() {
-                Ok(XmlEvent::Characters(s)) => return Ok(s),
-                Ok(event @ XmlEvent::EndElement { .. }) => {
-                    self.queued_event = Some(event);
-                    return Ok("".to_owned());
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
-                }
-                Ok(XmlEvent::StartElement { .. }) => {
-                    return Err(self.with_pos(ErrorKind::UnexpectedXmlOpeningTag));
-                }
-                Ok(XmlEvent::ProcessingInstruction { .. }) => (),
-                Ok(XmlEvent::StartDocument { .. })
-                | Ok(XmlEvent::CData(_))
-                | Ok(XmlEvent::Comment(_))
-                | Ok(XmlEvent::Whitespace(_)) => {
-                    unreachable!("parser does not output CData, Comment or Whitespace events");
-                }
-                Err(err) => return Err(from_xml_error(err)),
+impl From<XmlReaderError> for ErrorKind {
+    fn from(err: XmlReaderError) -> Self {
+        match err {
+            XmlReaderError::Io(err) if err.kind() == io::ErrorKind::UnexpectedEof => {
+                ErrorKind::UnexpectedEof
             }
+            XmlReaderError::Io(err) => {
+                let err = if let Some(code) = err.raw_os_error() {
```

Comment:
Can this not just be `XmlReaderError::Io(err) => ErrorKind::Io(err)`?

Reply:
apparently it can be, yeah. Some of the code that I didn't touch myself yet ^^